File: 2023/day05/solution.py
#!/usr/bin/env python3
import sys
import re
from collections import deque

DEBUG = 0
FILENAME = 'input.txt'
PART1 = PART2 = True

for arg in sys.argv[1:]:
  if arg.startswith('-e'): FILENAME = f'''example{arg[2:] if len(arg)>2 else ''}.txt'''
  elif arg.startswith('-d'): DEBUG = int(arg[2:]) if len(arg) > 2 else 1
  else:
    match arg:
      case '-i':
        FILENAME = 'input.txt'
      case '-p1':
        PART1, PART2 = True, False
      case '-p2':
        PART1, PART2 = False, True
      case '-p0':
        PART1, PART2 = False, False
      case _:
        raise Exception(f'unknown {arg=}')

with open(FILENAME, 'r') as f:
  text = f.read()
seeds = [int(x) for x in text.splitlines()[0].split(': ')[1].split()]
mappings = {}
section_regex = re.compile(r'^(\w+)-to-(\w+) map:((?:\n\d+ \d+ \d+)+)', re.MULTILINE)
line_regex = re.compile(r'^(\d+) (\d+) (\d+)$', re.MULTILINE)
for section in re.finditer(section_regex, text):
  if DEBUG > 1: print(f'{section.group(1)} -> {section.group(2)}')
  section_mappings = []
  for line in re.finditer(line_regex, section.group(3)):
    source_start = int(line.group(2))
    dest_start = int(line.group(1))
    length = int(line.group(3))
    section_mapping = {
      'start': source_start,
      'end': source_start + length - 1,
      'offset': dest_start - source_start,
    }
    section_mappings += [section_mapping]
    if DEBUG > 1: print(f'  {line.group(1)} {line.group(2)} {line.group(3)} {section_mapping=}')
  mappings[(section.group(1), section.group(2))] = sorted(section_mappings, key=lambda x: x['start'])
if DEBUG > 1: print(f'{mappings=}')
order = {x[0]: x[1] for x in mappings.keys()}
if DEBUG > 1: print(f'{order=}')


def part1():

  def convert(from_type: str, to_type: str, value: int) -> int:
    for mapping in mappings[(from_type, to_type)]:
      if value >= mapping['start'] and value <= mapping['end']:
        if DEBUG > 2: print(f'    found {mapping=}')
        return value + mapping['offset']
    return value

  min_val = None
  for seed in seeds:
    val = seed
    if DEBUG > 0: print(f'{seed=}')
    for k, v in order.items():
      val = convert(k, v, val)
      if DEBUG > 0: print(f'  {k}->{v} {val=}')
    if min_val is None or min_val > val: min_val = val
  print(f'part 1: {min_val}')


def part2():

  def convert_range(from_type: str, to_type: str, value_ranges: list[tuple[int, int]]) -> list[tuple[int, int]]:
    type_mappings = mappings[(from_type, to_type)]
    result: list[tuple[int, int]] = []
    queue: deque[tuple[int, int]] = deque(value_ranges)
    mapped = bool(len(queue))
    while len(queue):
      mapped = False
      from_value, to_value = queue.popleft()
      for mapping in type_mappings:
        if mapping['start']<=from_value<=mapping['end'] or \
          mapping['start']<=to_value<=mapping['end']:
          if DEBUG > 1: print(f'  {from_value=} {to_value=} {mapping=}')
          # push mapped ranges onto result
          mapped_from = (max(from_value, mapping['start']), min(to_value, mapping['end']))
          mapped_to = tuple(x + mapping['offset'] for x in mapped_from)
          result.append(mapped_to)  #type: ignore
          if DEBUG > 2: print(f'    mapped {mapped_from=} {mapped_to=}')
          # push unmapped ranges back onto queue
          if from_value < mapping['start']:
            requeue = (from_value, mapping['start'] - 1)
            queue.append(requeue)
            if DEBUG > 2: print(f'    requeued start {requeue}')
          if to_value > mapping['end']:
            requeue = (mapping['end'] + 1, to_value)
            queue.append(requeue)
            if DEBUG > 2: print(f'    requeued end {requeue}')
          # don't evaluate any further type mappings and fetch the next queue item
          mapped = True
          break
      # no matching mapping - append to result
      if not mapped:
        if DEBUG > 2: print(f'  unmapped {from_value=} {to_value=}')
        if DEBUG > 3: print(f'    {from_type=} {to_type=} {type_mappings=}')
        result.append((from_value, to_value))

    if DEBUG > 1:
      print(f'  {value_ranges=}')
      print(f'  {result=}')
    assert len(result)
    return result

  min_val = None
  for seed_start, length in zip(seeds[::2], seeds[1::2]):
    seed_end = seed_start + length - 1
    if DEBUG > 0: print(f'{seed_start=} {length=} {seed_end=}')
    vals = [(seed_start, seed_end)]
    for k, v in order.items():
      if DEBUG > 1: print(f'{k}->{v} {vals=}')
      vals = convert_range(k, v, vals)
    if DEBUG > 1: print(f'final {vals=}')
    seed_min = min(x[0] for x in vals)
    if min_val is None or min_val > seed_min: min_val = seed_min
  print(f'part 2: {min_val}')


if PART1: part1()
if PART2: part2()

File: 2023/day05/solution.ts
#!/usr/bin/env -S deno --allow-read

const Maths = Math;

const DEBUG = Deno.args.reduce((acc, item) => (item.startsWith('-d') ? Number(item.slice(2) || '1') : acc), 0);
const FILENAME = Deno.args.reduce(
  (acc, item) => (item == '-i' ? 'input.txt' : item.startsWith('-e') ? `example${item.slice(2)}.txt` : acc),
  'example.txt'
);
const [PART1, PART2] = Deno.args.reduce(
  (acc, item) => (item == '-p0' ? [false, false] : item == '-p1' ? [true, false] : item == '-p2' ? [false, true] : acc),
  [true, true]
);

console.log({ FILENAME, DEBUG, PART1, PART2 });

// deno-lint-ignore no-explicit-any
const debug = (level: number, ...data: any[]) => {
  if (DEBUG >= level) console.debug(...data);
};

const input = await Deno.readTextFile(FILENAME);

const parseInput = () => {
  interface Mapping {
    from: number;
    to: number;
    offset: number;
  }

  const sections = input.split('\n\n');

  const seeds = sections[0]
    .split(/:\s+/)[1]
    .split(/\s+/)
    .map((token) => Number(token));

  const conversions: Map<string, Mapping[]> = new Map(
    sections.slice(1).map((section) => [
      section.split(' ')[0],
      section
        .split('\n')
        .slice(1)
        .filter((line) => line.trim())
        .map((line) => {
          const tokens = line.split(/\s+/);
          return {
            // dest from, source from, count
            from: Number(tokens[1]),
            to: Number(tokens[1]) + Number(tokens[2]) - 1,
            offset: Number(tokens[0]) - Number(tokens[1]),
          };
        }),
    ])
  );

  return { seeds, conversions };
};

const part1 = () => {
  const { seeds, conversions } = parseInput();

  const convert = (keys: string[], value: number): number => {
    if (keys.length == 0) return value;
    const key = keys.shift()!;
    const mapping = conversions.get(key)!.find((i) => i.from <= value && value <= i.to);
    const result = typeof mapping != 'undefined' ? value + mapping.offset : value;
    debug(2, { key, value, result });
    return convert(keys, result);
  };

  const locations = seeds.map((seed) => convert([...conversions.keys()], seed));
  debug(1, { locations });
  const minLocation = Maths.min(...locations);

  console.log('part 1:', minLocation);
};

const part2 = () => {
  const { seeds, conversions } = parseInput();

  interface Range {
    from: number;
    to: number;
  }

  const convert = (keys: string[], ranges: Range[]): Range[] => {
    if (keys.length == 0) return ranges;
    const key = keys.shift()!;
    debug(1, { key, ranges });
    const queue = [...ranges];
    const results: Range[] = [];

    while (queue.length) {
      const range = queue.shift()!;
      // find the first mapping which at least partially covers our range
      const mapping = conversions.get(key)!.find((m) => m.from <= range.to && m.to >= range.from);
      if (typeof mapping != 'undefined') {
        const mapped: Range = {
          from: Maths.max(range.from, mapping.from) + mapping.offset,
          to: Maths.min(range.to, mapping.to) + mapping.offset,
        };
        const unmapped: Range[] = [];
        // push unmapped back to the queue
        if (mapping.from > range.from) unmapped.push({ from: range.from, to: mapping.from - 1 });
        if (mapping.to < range.to) unmapped.push({ from: mapping.to + 1, to: range.to });
        debug(2, { range, mapping, mapped, unmapped });
        results.push(mapped);
        queue.push(...unmapped);
      } else {
        debug(2, { range, mapping });
        results.push(range);
      }
    }

    const inputCount = ranges.map((r) => r.to - r.from + 1).reduce((acc, item) => acc + item, 0);
    const outputCount = results.map((r) => r.to - r.from + 1).reduce((acc, item) => acc + item, 0);
    debug(1, { key, ranges, results, inputCount, outputCount });
    if (inputCount != outputCount) throw new Error('bruh');

    return convert(keys, results);
  };

  const locations = seeds
    .filter((_, i) => i % 2 == 0)
    .map((seed, i) => [seed, seed + seeds[i * 2 + 1] - 1])
    .flatMap(([from, to]) => convert([...conversions.keys()], [{ from, to }]));

  debug(1, { locations });
  const minLocation = Maths.min(...locations.map((location) => location.from));
  console.log('part 2:', minLocation);
};

if (PART1) part1();
if (PART2) part2();
